IllegalStateException when reading empty byte array

## Bug Report
Insert a row with a empty byte array (not null) then attempt to read it:

```java.lang.AssertionError: expectation "assertNext" failed (expected: onNext(); actual: onError(java.lang.IllegalStateException: ByteBuf does not contain BYTEA hex format))

	at reactor.test.MessageFormatter.assertionError(MessageFormatter.java:115)
	at reactor.test.MessageFormatter.failPrefix(MessageFormatter.java:104)
	at reactor.test.MessageFormatter.fail(MessageFormatter.java:73)
	at reactor.test.MessageFormatter.failOptional(MessageFormatter.java:88)
	at reactor.test.DefaultStepVerifierBuilder.lambda$consumeNextWith$1(DefaultStepVerifierBuilder.java:268)
	at reactor.test.DefaultStepVerifierBuilder$SignalEvent.test(DefaultStepVerifierBuilder.java:2207)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onSignal(DefaultStepVerifierBuilder.java:1483)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onExpectation(DefaultStepVerifierBuilder.java:1431)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onError(DefaultStepVerifierBuilder.java:1091)
	at reactor.core.publisher.MonoFlatMapMany$FlatMapManyInner.onError(MonoFlatMapMany.java:247)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onError(Operators.java:1863)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:819)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:589)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:569)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.innerError(FluxFlatMap.java:840)
	at reactor.core.publisher.FluxFlatMap$FlatMapInner.onError(FluxFlatMap.java:979)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onError(FluxHandle.java:196)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:101)
	at reactor.core.publisher.FluxTakeUntil$TakeUntilPredicateSubscriber.onNext(FluxTakeUntil.java:77)
	at reactor.core.publisher.FluxFilterFuseable$FilterFuseableSubscriber.onNext(FluxFilterFuseable.java:112)
	at reactor.core.publisher.FluxContextStart$ContextStartSubscriber.onNext(FluxContextStart.java:103)
	at io.r2dbc.postgresql.util.FluxDiscardOnCancel$FluxDiscardOnCancelSubscriber.onNext(FluxDiscardOnCancel.java:86)
	at reactor.core.publisher.FluxCreate$BufferAsyncSink.drain(FluxCreate.java:793)
	at reactor.core.publisher.FluxCreate$BufferAsyncSink.next(FluxCreate.java:718)
	at reactor.core.publisher.FluxCreate$SerializedSink.next(FluxCreate.java:153)
	at io.r2dbc.postgresql.client.ReactorNettyClient$Conversation.emit(ReactorNettyClient.java:700)
	at io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.emit(ReactorNettyClient.java:888)
	at io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.onNext(ReactorNettyClient.java:822)
	at io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.onNext(ReactorNettyClient.java:731)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:112)
	at reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onNext(FluxPeekFuseable.java:845)
	at reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:213)
	at reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:213)
	at reactor.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:218)
	at reactor.netty.channel.FluxReceive.onInboundNext(FluxReceive.java:351)
	at reactor.netty.channel.ChannelOperations.onInboundNext(ChannelOperations.java:348)
	at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:89)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:326)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:313)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:427)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:281)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1422)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:931)
	at io.netty.channel.kqueue.AbstractKQueueStreamChannel$KQueueStreamUnsafe.readReady(AbstractKQueueStreamChannel.java:544)
	at io.netty.channel.kqueue.AbstractKQueueChannel$AbstractKQueueUnsafe.readReady(AbstractKQueueChannel.java:381)
	at io.netty.channel.kqueue.KQueueEventLoop.processReady(KQueueEventLoop.java:211)
	at io.netty.channel.kqueue.KQueueEventLoop.run(KQueueEventLoop.java:289)
	at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1050)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: java.lang.IllegalStateException: ByteBuf does not contain BYTEA hex format
		at io.r2dbc.postgresql.codec.AbstractBinaryCodec.decode(AbstractBinaryCodec.java:66)
		at io.r2dbc.postgresql.codec.BinaryByteBufferCodec.doDecode(BinaryByteBufferCodec.java:40)
		at io.r2dbc.postgresql.codec.BinaryByteBufferCodec.doDecode(BinaryByteBufferCodec.java:32)
		at io.r2dbc.postgresql.codec.AbstractCodec.decode(AbstractCodec.java:82)
		at io.r2dbc.postgresql.codec.DefaultCodecs.decode(DefaultCodecs.java:128)
		at io.r2dbc.postgresql.PostgresqlRow.decode(PostgresqlRow.java:94)
		at io.r2dbc.postgresql.PostgresqlRow.get(PostgresqlRow.java:80)
		at io.r2dbc.postgresql.AbstractCodecIntegrationTests.lambda$testCodec$25(AbstractCodecIntegrationTests.java:480)
		at io.r2dbc.postgresql.PostgresqlResult.lambda$map$1(PostgresqlResult.java:112)
		at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:96)
		... 41 more
```

#### Versions

- Driver: master
- Database: Postgresql
- Java:
- OS: Ubuntu

#### Current Behavior

<!-- A clear and concise description of the behavior.-->

<details>
<summary>Stack trace</summary>

```
// your stack trace here
```

</details>
 
#### Table schema

<!--- Provide the table schema. -->

<details>
<summary>Input Code</summary>

```sql
-- your SQL here;
```

</details>

#### Steps to reproduce

<!-- Java/Kotlin/Scala/Groovy/… or Repo link to a Minimal, Reproducible Example if applicable.
Preferably raw R2DBC code without any 3rd-party dependencies. -->

<details>
<summary>Input Code</summary>

```java
// your code here;
```

</details>

#### Expected behavior/code

<!-- A clear and concise description of what you expected to happen (or code). -->

#### Possible Solution

<!-- Only if you have suggestions on a fix for the bug -->

#### Additional context

<!-- Add any other context about the problem here. Do not add code as screenshots. -->

--- a/src/main/java/io/r2dbc/postgresql/codec/AbstractBinaryCodec.java
+++ b/src/main/java/io/r2dbc/postgresql/codec/AbstractBinaryCodec.java
@@ -26,6 +26,7 @@ import io.r2dbc.postgresql.util.Assert;
 import io.r2dbc.postgresql.util.ByteBufUtils;
 
 import java.nio.charset.StandardCharsets;
+import java.util.Objects;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
@@ -34,7 +35,7 @@ import static io.r2dbc.postgresql.type.PostgresqlObjectId.BYTEA;
 
 abstract class AbstractBinaryCodec<T> extends AbstractCodec<T> {
 
-    private static final Pattern BLOB_PATTERN = Pattern.compile("\\\\x([\\p{XDigit}]+)");
+    private static final Pattern BLOB_PATTERN = Pattern.compile("\\\\x([\\p{XDigit}]+)?");
 
     private final ByteBufAllocator byteBufAllocator;
 
@@ -64,7 +65,9 @@ abstract class AbstractBinaryCodec<T> extends AbstractCodec<T> {
             if (!matcher.find()) {
                 throw new IllegalStateException("ByteBuf does not contain BYTEA hex format");
             }
-            decoded = ByteBufUtil.decodeHexDump(matcher.group(1));
+
+            String bytesHex = Objects.toString(matcher.group(1), "");
+            decoded = ByteBufUtil.decodeHexDump(bytesHex);
         } else {
             decoded = new byte[byteBuf.readableBytes()];
             byteBuf.readBytes(decoded);

--- a/src/main/java/io/r2dbc/postgresql/codec/BlobCodec.java
+++ b/src/main/java/io/r2dbc/postgresql/codec/BlobCodec.java
@@ -32,6 +32,7 @@ import reactor.util.annotation.Nullable;
 
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
+import java.util.Objects;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
@@ -99,7 +100,7 @@ final class BlobCodec extends AbstractCodec<Blob> {
 
     private static final class ByteABlob implements Blob {
 
-        private static final Pattern BLOB_PATTERN = Pattern.compile("\\\\x([\\p{XDigit}]+)");
+        private static final Pattern BLOB_PATTERN = Pattern.compile("\\\\x([\\p{XDigit}]+)?");
 
         private final ByteBuf byteBuf;
 
@@ -132,7 +133,8 @@ final class BlobCodec extends AbstractCodec<Blob> {
                     throw new IllegalStateException("ByteBuf does not contain BYTEA hex format");
                 }
 
-                return ByteBuffer.wrap(ByteBufUtil.decodeHexDump(matcher.group(1)));
+                String bytesHex = Objects.toString(matcher.group(1), "");
+                return ByteBuffer.wrap(ByteBufUtil.decodeHexDump(bytesHex));
             }).doAfterTerminate(this.byteBuf::release);
         }
     }

--- a/src/test/java/io/r2dbc/postgresql/AbstractCodecIntegrationTests.java
+++ b/src/test/java/io/r2dbc/postgresql/AbstractCodecIntegrationTests.java
@@ -100,6 +100,7 @@ abstract class AbstractCodecIntegrationTests extends AbstractIntegrationTests {
     @Test
     void binary() {
         testCodec(ByteBuffer.class, ByteBuffer.wrap(new byte[]{1, 2, 3, 4}), "BYTEA");
+        testCodec(ByteBuffer.class, ByteBuffer.wrap(new byte[]{}), "BYTEA");
     }
 
     @Test
@@ -115,28 +116,27 @@ abstract class AbstractCodecIntegrationTests extends AbstractIntegrationTests {
 
     @Test
     void blob() {
-        byte[] bytes = {1, 2, 3, 4};
-        testCodec(Blob.class,
-            new Blob() {
-
-                @Override
-                public Publisher<Void> discard() {
-                    return Mono.empty();
-                }
-
-                @Override
-                public Publisher<ByteBuffer> stream() {
-                    return Mono.just(ByteBuffer.wrap(bytes));
-                }
-            },
-            (actual, expected) -> Flux.zip(
+        BiConsumer<Blob, Blob> equality = (actual, expected) -> Flux.zip(
                 Flux.from(actual.stream()).reduce(TEST.heapBuffer(), ByteBuf::writeBytes),
                 Flux.from(expected.stream()).reduce(TEST.heapBuffer(), ByteBuf::writeBytes)
-            )
+        )
                 .as(StepVerifier::create)
                 .assertNext(t -> assertThat(t.getT1()).isEqualTo(t.getT2()))
-                .verifyComplete()
-            , "BYTEA");
+                .verifyComplete();
+
+        Function<byte[], Blob> byteToBlob = (bytes) -> new Blob() {
+            @Override
+            public Publisher<Void> discard() {
+                return Mono.empty();
+            }
+
+            @Override
+            public Publisher<ByteBuffer> stream() {
+                return Mono.just(ByteBuffer.wrap(bytes));
+            }};
+
+        testCodec(Blob.class, byteToBlob.apply(new byte[]{1, 2, 3, 4}), equality, "BYTEA");
+        testCodec(Blob.class, byteToBlob.apply(new byte[]{}), equality, "BYTEA");
     }
 
     @Test

